src/pal/inc/unixasmmacrosarm.inc

```diff
@@ -167,24 +168,43 @@ C_FUNC(\Name\()_End):
 
 .macro EPILOG_WITH_TRANSITION_BLOCK_RETURN
 
-        free_stack __PWTB_StackAlloc
-        POP_CALLEE_SAVED_REGISTERS
-        free_stack 16
-        bx lr
+	// Do not restore r7, lr until EPILOG.
+
+	add r6, sp, #(__PWTB_StackAlloc)
+	ldm r6, {r4-r5}
+
+	add r6, sp, #((__PWTB_StackAlloc) + 4 * 4)
+	ldm r6, {r8-r11}
+
+	ldr r6, [sp, #((__PWTB_StackAlloc) + 8)]
+
+	free_stack ((__PWTB_StackAlloc) + 4 * 4 + 9 * 4)
+	EPILOG_POP "{r7, pc}"
 
 .endm
 
 .macro EPILOG_WITH_TRANSITION_BLOCK_TAILCALL
 
         .if (__PWTB_SaveFPArgs == 1)
-                add r6, sp, #(__PWTB_FloatArgumentRegisters)
+                mov r6, sp
                 vldm r6, {d0-d7}
         .endif
 
-        free_stack __PWTB_StackAlloc
-        POP_CALLEE_SAVED_REGISTERS
-        POP_ARGUMENT_REGISTERS
-        
+	// Do not restore r7, lr until EPILOG.
+
+	add r6, sp, #((__PWTB_StackAlloc) + 9 * 4)
+	ldm r6, {r0-r3}
+
+	add r6, sp, #(__PWTB_StackAlloc)
+	ldm r6, {r4-r5}
+
+	add r6, sp, #((__PWTB_StackAlloc) + 4 * 4)
+	ldm r6, {r8-r11}
+
+	ldr r6, [sp, #((__PWTB_StackAlloc) + 8)]
+
+	free_stack ((__PWTB_StackAlloc) + 4 * 4 + 9 * 4)
+	EPILOG_POP "{r7, lr}"
```

Comment:
Same happens with me as well.

`pop {r7, lr}` makes it lose the current context as we do longer have valid FP (r7).
Unless we have an atomic operation to do `pop {r7}` and `mov pc r12`, we are not going to be safe here.

We may overwrite the saved LR value of stack and use `pop {r7, pc}` not `pop {r7, lr}`. However, unless the value of this stack is not going to be used ever again, we shouldn't do that. I'll consider this option as well.